nexus/src/db/datastore.rs

```diff
@@ -282,10 +310,42 @@ impl DataStore {
                     LookupType::ByName(name.as_str().to_owned()),
                 )
             })
+            .map(|o| (authz::FLEET.organization(o.id()), o))
+    }
+
+    /// Look up the id for an organization based on its name
+    ///
+    /// Returns an [`authz::Organization`] (which makes the id available).
+    ///
+    /// This function does no authz checks because it is not possible to know
+    /// just by looking up an Organization's id what privileges are required.
+    pub async fn organization_lookup_id(
+        &self,
+        name: &Name,
+    ) -> LookupResult<authz::Organization> {
+        self.organization_lookup_noauthz(name).await.map(|(o, _)| o)
+    }
+
+    /// Lookup an organization by name.
+    pub async fn organization_fetch(
+        &self,
+        opctx: &OpContext,
+        name: &Name,
+    ) -> LookupResult<(authz::Organization, Organization)> {
+        let (authz_org, db_org) =
+            self.organization_lookup_noauthz(name).await?;
```

Comment:
I'm totally on-board with the "noauthz" prefix, but is it necessary to use both "fetch" and "lookup" as different verbs to access the organization?

We have:
- Get an Organization, without authZ
- Get an Organization ID, without authZ
- Get an Organization, with authZ

Couldn't these be:
- organization_lookup_noauthz
- organization_lookup_id
- organization_lookup (the "authz") is implied

?

I know this is kinda a nit, but given the large surface area of the DB, it seems like an as-simple-as-possible naming convention will be worthwhile investing into.

Reply:
Ha, I think maybe what you find clearer is what I find more confusing.  If it's okay, I'll think about this some more as I iterate on the other lookup functions.  I expect these to evolve, potentially a lot: at the very least, we're going to be adding lookup-by-id, for example.

Reply:
Narrowing my question, when do you view "fetch" as an appropriate verb vs "lookup"?

Reply:
I'll explain my intuition below, with the caveat that I don't think my intuition here is necessarily based on a good reason.  It's probably a product of the systems I've worked on before.  I can see how a reader might not have the same intuition and I'm on board for changing the names.  I'm just not sure how to do it yet and I don't think it makes sense to spend a lot of time on it right now because I think a bunch of upcoming changes are likely to invalidate whatever convention we try to establish now based only on these three functions.

----

"fetch" to me suggests:

- a public interface (because it's a high-level-sounding operation)
- that it's authorization-protected (since it's a public interface)
- that it accepts as input whatever unique identifier the caller uses to refer to objects
- that it returns the whole record

"lookup" to me is much more general.  It could be an internal or external function.  I wouldn't have expectations about whether authorization is done.  It might accept any kind of criteria (not even necessarily an identifier) and return any number of objects or even some field from some object.

I don't like the specific thing you suggested because the function names look so similar that I think I'd find it hard to keep track of which ones do authz.  (After all, you said the "authz" is implied in `organization_lookup`, but it's not implied in `organization_lookup_by_id`, even though it doesn't have `_noauthz`.)

Reply:
For the record, the first time I read through your explanation, I found it strange. Now it makes a lot of sense to me. Not sure what that means, but I find the distinction you draw—between auth-gated functions that return a whole resource and internal non-auth-gated functions that return IDs and/or auth metadata about a resource—to be a coherent one, so I think assigning the two sides names make sense. We may have to get over some prior intuitions about what these words mean but that's ok. I have a feeling `lookup` and `fetch` are about as good as it's going to get. We have the same problem coming up with words for "thing": object, resource, model, row, data, etc.